run.go

```diff
@@ -86,16 +120,142 @@ func createGrpcServer() *grpc.Server {
 	return service.GRPCServer
 }
 
-func startPrometheusServer() {
+func startPrometheusServer(started func()) {
 	service.PrometheusServer = &http.Server{Addr: service.PrometheusConfig.Address()}
 
 	http.Handle("/metrics", promhttp.Handler())
 	logrus.Infof("Prometheus metrics at http://%s/metrics", service.PrometheusConfig.Address())
 
+	service.prometheusRunning = true
+	started()
+
 	go func() {
 		if err := service.PrometheusServer.ListenAndServe(); err != nil {
 			// cannot panic, because this probably is an intentional close
 			logrus.Errorf("Prometheus http server: ListenAndServe() error: %s", err)
 		}
 	}()
 }
+
+func startProbeServer() {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/startz", startzHandler)
+	mux.HandleFunc("/readyz", readyzHandler)
+	mux.HandleFunc("/alivez", alivezHandler)
+
+	service.ProbeServer = &http.Server{
+		Addr:    service.ProbeConfig.Address(),
+		Handler: mux,
+	}
+
+	logrus.Infof("Serving health probes on %s [/startz, /readyz, /alivez]", service.ProbeConfig.Address())
+	go func() {
+		if err := service.ProbeServer.ListenAndServe(); err != nil {
+			// cannot panic, because this probably is an intentional close
+			logrus.Errorf("Probe http server: ListenAndServe() error: %s", err)
+		}
+	}()
+}
+
+func startzHandler(writer http.ResponseWriter, request *http.Request) {
+	pp := service.probes
+	if len(pp) == 0 {
+		return
+	}
+	var nonStartedServices []string
+	for _, p := range pp {
+		if !p.started {
+			nonStartedServices = append(nonStartedServices, p.name)
+		}
+	}
+	if len(nonStartedServices) == 0 {
+		return
+	}
+	writer.WriteHeader(503)
+	msg := fmt.Sprintf("[%s] not started yet", strings.Join(nonStartedServices, ", "))
+	logrus.Warningf("startz probe failed: %s", msg)
+	_, err := writer.Write([]byte(msg))
+	if err != nil {
+		logrus.Errorf("unable to write failed startz probe response : %v", err)
+	}
+}
+
+func readyzHandler(writer http.ResponseWriter, request *http.Request) {
+	pp := service.probes
+	if len(pp) == 0 {
+		return
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(len(pp))
+	errs := make([]error, len(pp))
+	ctx := request.Context()
+	for i, p := range pp {
+		go func(i int, p *probe) {
+			err := p.ready(ctx)
+			if err != nil {
+				errs[i] = err
+			} else if ctx.Err() != nil {
+				errs[i] = err
+			}
+			wg.Done()
+		}(i, p)
+	}
+	wg.Wait()
+	var errMsgs []string
+	for i, e := range errs {
+		if e == nil {
+			continue
+		}
+		errMsgs = append(errMsgs, fmt.Sprintf("\t%s:\n\t\t%v", pp[i].name, e))
+	}
+	if len(errMsgs) == 0 {
+		return
+	}
+	writer.WriteHeader(503)
+	msg := strings.Join(errMsgs, "\n")
+	logrus.Errorf("readyz probe failed:\n%s", msg)
+	_, err := writer.Write([]byte(msg))
+	if err != nil {
+		logrus.Errorf("unable to write failed readyz probe response : %v", err)
+	}
+}
+
+func alivezHandler(writer http.ResponseWriter, request *http.Request) {
```

Comment:
These probe handlers have a lot of duplication between then, can some logic be extracted into dedicated funcs?

Reply:
It can, on it 🙌

Reply:
done